File: backend/backup/feature-flag-service/go/internal/repository/redis_feature_flag_repository.go
package repository

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"
	"sync"
	"time"

	"github.com/codebridge/feature-flag-service/internal/model"
	"github.com/go-redis/redis/v8"
)

// RedisFeatureFlagRepository implements FeatureFlagRepository using Redis
type RedisFeatureFlagRepository struct {
	client       *redis.Client
	ttlSeconds   int
	pubSubClient *redis.PubSub
	mutex        sync.Mutex
	callbacks    []func(*model.FeatureFlag)
}

// NewRedisFeatureFlagRepository creates a new Redis-backed feature flag repository
func NewRedisFeatureFlagRepository(client *redis.Client, ttlSeconds int) *RedisFeatureFlagRepository {
	return &RedisFeatureFlagRepository{
		client:     client,
		ttlSeconds: ttlSeconds,
		callbacks:  make([]func(*model.FeatureFlag), 0),
	}
}

// GetFlag gets a flag by key and namespace
func (r *RedisFeatureFlagRepository) GetFlag(ctx context.Context, key, namespace string) (*model.FeatureFlag, error) {
	// Construct the Redis key
	redisKey := getFlagKey(key, namespace)

	// Get the flag from Redis
	data, err := r.client.Get(ctx, redisKey).Result()
	if err != nil {
		if err == redis.Nil {
			return nil, nil // Flag not found
		}
		return nil, fmt.Errorf("failed to get flag from Redis: %w", err)
	}

	// Unmarshal the flag
	var flag model.FeatureFlag
	if err := json.Unmarshal([]byte(data), &flag); err != nil {
		return nil, fmt.Errorf("failed to unmarshal flag: %w", err)
	}

	return &flag, nil
}

// GetFlags gets multiple flags by keys and namespace
func (r *RedisFeatureFlagRepository) GetFlags(ctx context.Context, keys []string, namespace string) ([]*model.FeatureFlag, error) {
	if len(keys) == 0 {
		return make([]*model.FeatureFlag, 0), nil
	}

	// Construct Redis keys
	redisKeys := make([]string, len(keys))
	for i, key := range keys {
		redisKeys[i] = getFlagKey(key, namespace)
	}

	// Get flags from Redis
	data, err := r.client.MGet(ctx, redisKeys...).Result()
	if err != nil {
		return nil, fmt.Errorf("failed to get flags from Redis: %w", err)
	}

	// Unmarshal flags
	flags := make([]*model.FeatureFlag, 0, len(data))
	for i, item := range data {
		if item == nil {
			continue // Flag not found
		}

		var flag model.FeatureFlag
		if err := json.Unmarshal([]byte(item.(string)), &flag); err != nil {
			return nil, fmt.Errorf("failed to unmarshal flag: %w", err)
		}

		flags = append(flags, &flag)
	}

	return flags, nil
}

// SetFlag sets a flag
func (r *RedisFeatureFlagRepository) SetFlag(ctx context.Context, flag *model.FeatureFlag) error {
	// Update timestamps
	flag.UpdatedAt = time.Now()

	// Marshal the flag
	data, err := json.Marshal(flag)
	if err != nil {
		return fmt.Errorf("failed to marshal flag: %w", err)
	}

	// Construct the Redis key
	redisKey := getFlagKey(flag.Key, flag.Namespace)

	// Set the flag in Redis
	var ttl time.Duration
	if flag.Temporary && flag.ExpiresAt != nil {
		// Use the flag's expiration time
		ttl = time.Until(*flag.ExpiresAt)
	} else if r.ttlSeconds > 0 {
		// Use the repository's TTL
		ttl = time.Duration(r.ttlSeconds) * time.Second
	} else {
		// No TTL
		ttl = 0
	}

	if err := r.client.Set(ctx, redisKey, data, ttl).Err(); err != nil {
		return fmt.Errorf("failed to set flag in Redis: %w", err)
	}

	// Add to index
	if err := r.addToIndex(ctx, flag); err != nil {
		return fmt.Errorf("failed to add flag to index: %w", err)
	}

	// Publish update
	if err := r.publishUpdate(ctx, flag); err != nil {
		return fmt.Errorf("failed to publish flag update: %w", err)
	}

	return nil
}

// DeleteFlag deletes a flag
func (r *RedisFeatureFlagRepository) DeleteFlag(ctx context.Context, key, namespace string) error {
	// Get the flag first to remove from indexes
	flag, err := r.GetFlag(ctx, key, namespace)
	if err != nil {
		return fmt.Errorf("failed to get flag for deletion: %w", err)
	}

	if flag == nil {
		return nil // Flag not found, nothing to delete
	}

	// Construct the Redis key
	redisKey := getFlagKey(key, namespace)

	// Delete the flag from Redis
	if err := r.client.Del(ctx, redisKey).Err(); err != nil {
		return fmt.Errorf("failed to delete flag from Redis: %w", err)
	}

	// Remove from index
	if err := r.removeFromIndex(ctx, flag); err != nil {
		return fmt.Errorf("failed to remove flag from index: %w", err)
	}

	// Publish deletion
	flag.Value = nil // Clear the value to indicate deletion
	if err := r.publishUpdate(ctx, flag); err != nil {
		return fmt.Errorf("failed to publish flag deletion: %w", err)
	}

	return nil
}

// ListFlags lists flags by namespace and optional prefix
func (r *RedisFeatureFlagRepository) ListFlags(ctx context.Context, namespace, prefix string) ([]*model.FeatureFlag, error) {
	// Construct the Redis key pattern
	pattern := getNamespaceKey(namespace) + ":*"
	if prefix != "" {
		pattern = getNamespaceKey(namespace) + ":" + prefix + "*"
	}

	// Scan for matching keys
	var cursor uint64
	var keys []string
	var err error

	for {
		var batch []string
		cursor, batch, err = r.client.Scan(ctx, cursor, pattern, 100).Result()
		if err != nil {
			return nil, fmt.Errorf("failed to scan Redis keys: %w", err)
		}

		keys = append(keys, batch...)

		if cursor == 0 {
			break
		}
	}

	if len(keys) == 0 {
		return make([]*model.FeatureFlag, 0), nil
	}

	// Get flags from Redis
	data, err := r.client.MGet(ctx, keys...).Result()
	if err != nil {
		return nil, fmt.Errorf("failed to get flags from Redis: %w", err)
	}

	// Unmarshal flags
	flags := make([]*model.FeatureFlag, 0, len(data))
	for i, item := range data {
		if item == nil {
			continue // Flag not found
		}

		var flag model.FeatureFlag
		if err := json.Unmarshal([]byte(item.(string)), &flag); err != nil {
			return nil, fmt.Errorf("failed to unmarshal flag: %w", err)
		}

		flags = append(flags, &flag)
	}

	return flags, nil
}

// ListFlagsByTag lists flags by tag
func (r *RedisFeatureFlagRepository) ListFlagsByTag(ctx context.Context, namespace, tagKey, tagValue string) ([]*model.FeatureFlag, error) {
	// Construct the Redis key for the tag index
	indexKey := getTagIndexKey(namespace, tagKey, tagValue)

	// Get flag keys from the index
	flagKeys, err := r.client.SMembers(ctx, indexKey).Result()
	if err != nil {
		return nil, fmt.Errorf("failed to get flag keys from tag index: %w", err)
	}

	if len(flagKeys) == 0 {
		return make([]*model.FeatureFlag, 0), nil
	}

	// Get flags from Redis
	data, err := r.client.MGet(ctx, flagKeys...).Result()
	if err != nil {
		return nil, fmt.Errorf("failed to get flags from Redis: %w", err)
	}

	// Unmarshal flags
	flags := make([]*model.FeatureFlag, 0, len(data))
	for i, item := range data {
		if item == nil {
			continue // Flag not found
		}

		var flag model.FeatureFlag
		if err := json.Unmarshal([]byte(item.(string)), &flag); err != nil {
			return nil, fmt.Errorf("failed to unmarshal flag: %w", err)
		}

		flags = append(flags, &flag)
	}

	return flags, nil
}

// Subscribe subscribes to flag updates
func (r *RedisFeatureFlagRepository) Subscribe(ctx context.Context, callback func(*model.FeatureFlag)) error {
	r.mutex.Lock()
	defer r.mutex.Unlock()

	// Add the callback
	r.callbacks = append(r.callbacks, callback)

	// Start the subscription if not already started
	if r.pubSubClient == nil {
		pubSub := r.client.Subscribe(ctx, "feature-flag-updates")
		r.pubSubClient = pubSub

		// Start a goroutine to handle messages
		go r.handleMessages(ctx)
	}

	return nil
}

// Unsubscribe unsubscribes from flag updates
func (r *RedisFeatureFlagRepository) Unsubscribe(ctx context.Context) error {
	r.mutex.Lock()
	defer r.mutex.Unlock()

	// Clear callbacks
	r.callbacks = make([]func(*model.FeatureFlag), 0)

	// Close the subscription if it exists
	if r.pubSubClient != nil {
		if err := r.pubSubClient.Close(); err != nil {
			return fmt.Errorf("failed to close Redis PubSub: %w", err)
		}
		r.pubSubClient = nil
	}

	return nil
}

// handleMessages handles messages from the Redis PubSub
func (r *RedisFeatureFlagRepository) handleMessages(ctx context.Context) {
	for {
		msg, err := r.pubSubClient.ReceiveMessage(ctx)
		if err != nil {
			// Check if the context is done
			if ctx.Err() != nil {
				return
			}

			// Log the error and continue
			fmt.Printf("Error receiving message from Redis PubSub: %v\n", err)
			continue
		}

		// Unmarshal the flag
		var flag model.FeatureFlag
		if err := json.Unmarshal([]byte(msg.Payload), &flag); err != nil {
			fmt.Printf("Failed to unmarshal flag from Redis PubSub: %v\n", err)
			continue
		}

		// Call the callbacks
		r.mutex.Lock()
		for _, callback := range r.callbacks {
			callback(&flag)
		}
		r.mutex.Unlock()
	}
}

// addToIndex adds a flag to the indexes
func (r *RedisFeatureFlagRepository) addToIndex(ctx context.Context, flag *model.FeatureFlag) error {
	// Add to namespace index
	namespaceKey := getNamespaceIndexKey(flag.Namespace)
	if err := r.client.SAdd(ctx, namespaceKey, flag.Key).Err(); err != nil {
		return fmt.Errorf("failed to add flag to namespace index: %w", err)
	}

	// Add to tag indexes
	for tagKey, tagValue := range flag.Tags {
		tagIndexKey := getTagIndexKey(flag.Namespace, tagKey, tagValue)
		flagKey := getFlagKey(flag.Key, flag.Namespace)
		if err := r.client.SAdd(ctx, tagIndexKey, flagKey).Err(); err != nil {
			return fmt.Errorf("failed to add flag to tag index: %w", err)
		}
	}

	return nil
}

// removeFromIndex removes a flag from the indexes
func (r *RedisFeatureFlagRepository) removeFromIndex(ctx context.Context, flag *model.FeatureFlag) error {
	// Remove from namespace index
	namespaceKey := getNamespaceIndexKey(flag.Namespace)
	if err := r.client.SRem(ctx, namespaceKey, flag.Key).Err(); err != nil {
		return fmt.Errorf("failed to remove flag from namespace index: %w", err)
	}

	// Remove from tag indexes
	for tagKey, tagValue := range flag.Tags {
		tagIndexKey := getTagIndexKey(flag.Namespace, tagKey, tagValue)
		flagKey := getFlagKey(flag.Key, flag.Namespace)
		if err := r.client.SRem(ctx, tagIndexKey, flagKey).Err(); err != nil {
			return fmt.Errorf("failed to remove flag from tag index: %w", err)
		}
	}

	return nil
}

// publishUpdate publishes a flag update
func (r *RedisFeatureFlagRepository) publishUpdate(ctx context.Context, flag *model.FeatureFlag) error {
	// Marshal the flag
	data, err := json.Marshal(flag)
	if err != nil {
		return fmt.Errorf("failed to marshal flag for publishing: %w", err)
	}

	// Publish the update
	if err := r.client.Publish(ctx, "feature-flag-updates", data).Err(); err != nil {
		return fmt.Errorf("failed to publish flag update: %w", err)
	}

	return nil
}

// Helper functions for Redis keys

// getFlagKey returns the Redis key for a flag
func getFlagKey(key, namespace string) string {
	return fmt.Sprintf("feature-flag:%s:%s", namespace, key)
}

// getNamespaceKey returns the Redis key prefix for a namespace
func getNamespaceKey(namespace string) string {
	return fmt.Sprintf("feature-flag:%s", namespace)
}

// getNamespaceIndexKey returns the Redis key for a namespace index
func getNamespaceIndexKey(namespace string) string {
	return fmt.Sprintf("feature-flag:index:namespace:%s", namespace)
}

// getTagIndexKey returns the Redis key for a tag index
func getTagIndexKey(namespace, tagKey, tagValue string) string {
	return fmt.Sprintf("feature-flag:index:tag:%s:%s:%s", namespace, tagKey, tagValue)
}

File: workspace/backend/backup/feature-flag-service/python/feature_flag_service/repository/redis_feature_flag_repository.py
"""
Redis implementation of the feature flag repository.
"""

import asyncio
import json
import logging
from datetime import datetime
from typing import Callable, Dict, List, Optional, Set

import redis.asyncio as redis

from feature_flag_service.model.feature_flag import FeatureFlag
from feature_flag_service.repository.feature_flag_repository import FeatureFlagRepository

logger = logging.getLogger(__name__)


class RedisFeatureFlagRepository(FeatureFlagRepository):
    """
    Redis implementation of the feature flag repository.
    """

    def __init__(
        self,
        host: str = "localhost",
        port: int = 6379,
        password: Optional[str] = None,
        db: int = 0,
        ttl_seconds: int = 0,
    ):
        """
        Initialize the repository.
        """
        self.redis = redis.Redis(
            host=host,
            port=port,
            password=password,
            db=db,
            decode_responses=True,
        )
        self.ttl_seconds = ttl_seconds
        self.pubsub = None
        self.callbacks: Set[Callable[[FeatureFlag], None]] = set()
        self.pubsub_task = None

    async def get_flag(self, key: str, namespace: str) -> Optional[FeatureFlag]:
        """
        Get a flag by key and namespace.
        """
        redis_key = self._get_flag_key(key, namespace)
        data = await self.redis.get(redis_key)

        if not data:
            return None

        try:
            flag_dict = json.loads(data)
            return FeatureFlag.parse_obj(flag_dict)
        except Exception as e:
            logger.error(f"Failed to parse flag data: {e}")
            return None

    async def get_flags(self, keys: List[str], namespace: str) -> List[FeatureFlag]:
        """
        Get multiple flags by keys and namespace.
        """
        if not keys:
            return []

        # Construct Redis keys
        redis_keys = [self._get_flag_key(key, namespace) for key in keys]

        # Get flags from Redis
        data_list = await self.redis.mget(redis_keys)

        # Parse flags
        flags = []
        for data in data_list:
            if not data:
                continue

            try:
                flag_dict = json.loads(data)
                flags.append(FeatureFlag.parse_obj(flag_dict))
            except Exception as e:
                logger.error(f"Failed to parse flag data: {e}")

        return flags

    async def set_flag(self, flag: FeatureFlag) -> None:
        """
        Set a flag.
        """
        # Update timestamps
        flag.updated_at = datetime.now()

        # Serialize the flag
        flag_dict = flag.dict()
        data = json.dumps(flag_dict)

        # Construct the Redis key
        redis_key = self._get_flag_key(flag.key, flag.namespace)

        # Set the flag in Redis
        if flag.temporary and flag.expires_at:
            # Use the flag's expiration time
            ttl = int((flag.expires_at - datetime.now()).total_seconds())
            if ttl <= 0:
                # Flag is already expired
                return
            await self.redis.setex(redis_key, ttl, data)
        elif self.ttl_seconds > 0:
            # Use the repository's TTL
            await self.redis.setex(redis_key, self.ttl_seconds, data)
        else:
            # No TTL
            await self.redis.set(redis_key, data)

        # Add to index
        await self._add_to_index(flag)

        # Publish update
        await self._publish_update(flag)

    async def delete_flag(self, key: str, namespace: str) -> None:
        """
        Delete a flag.
        """
        # Get the flag first to remove from indexes
        flag = await self.get_flag(key, namespace)
        if not flag:
            return

        # Construct the Redis key
        redis_key = self._get_flag_key(key, namespace)

        # Delete the flag from Redis
        await self.redis.delete(redis_key)

        # Remove from index
        await self._remove_from_index(flag)

        # Publish deletion
        flag.value = None  # Clear the value to indicate deletion
        await self._publish_update(flag)

    async def list_flags(self, namespace: str, prefix: Optional[str] = None) -> List[FeatureFlag]:
        """
        List flags by namespace and optional prefix.
        """
        # Construct the Redis key pattern
        pattern = f"{self._get_namespace_key(namespace)}:*"
        if prefix:
            pattern = f"{self._get_namespace_key(namespace)}:{prefix}*"

        # Scan for matching keys
        cursor = 0
        keys = []

        while True:
            cursor, batch = await self.redis.scan(cursor, match=pattern, count=100)
            keys.extend(batch)

            if cursor == 0:
                break

        if not keys:
            return []

        # Get flags from Redis
        data_list = await self.redis.mget(keys)

        # Parse flags
        flags = []
        for data in data_list:
            if not data:
                continue

            try:
                flag_dict = json.loads(data)
                flags.append(FeatureFlag.parse_obj(flag_dict))
            except Exception as e:
                logger.error(f"Failed to parse flag data: {e}")

        return flags

    async def list_flags_by_tag(self, namespace: str, tag_key: str, tag_value: str) -> List[FeatureFlag]:
        """
        List flags by tag.
        """
        # Construct the Redis key for the tag index
        index_key = self._get_tag_index_key(namespace, tag_key, tag_value)

        # Get flag keys from the index
        flag_keys = await self.redis.smembers(index_key)

        if not flag_keys:
            return []

        # Get flags from Redis
        data_list = await self.redis.mget(flag_keys)

        # Parse flags
        flags = []
        for data in data_list:
            if not data:
                continue

            try:
                flag_dict = json.loads(data)
                flags.append(FeatureFlag.parse_obj(flag_dict))
            except Exception as e:
                logger.error(f"Failed to parse flag data: {e}")

        return flags

    async def subscribe(self, callback: Callable[[FeatureFlag], None]) -> None:
        """
        Subscribe to flag updates.
        """
        # Add the callback
        self.callbacks.add(callback)

        # Start the subscription if not already started
        if not self.pubsub:
            self.pubsub = self.redis.pubsub()
            await self.pubsub.subscribe("feature-flag-updates")

            # Start a task to handle messages
            if not self.pubsub_task:
                self.pubsub_task = asyncio.create_task(self._handle_messages())

    async def unsubscribe(self) -> None:
        """
        Unsubscribe from flag updates.
        """
        # Clear callbacks
        self.callbacks.clear()

        # Close the subscription if it exists
        if self.pubsub:
            await self.pubsub.unsubscribe()
            self.pubsub = None

        # Cancel the task if it exists
        if self.pubsub_task:
            self.pubsub_task.cancel()
            self.pubsub_task = None

    async def _handle_messages(self) -> None:
        """
        Handle messages from the Redis PubSub.
        """
        try:
            async for message in self.pubsub.listen():
                if message["type"] != "message":
                    continue

                try:
                    # Parse the flag
                    flag_dict = json.loads(message["data"])
                    flag = FeatureFlag.parse_obj(flag_dict)

                    # Call the callbacks
                    for callback in self.callbacks:
                        try:
                            callback(flag)
                        except Exception as e:
                            logger.error(f"Error in flag update callback: {e}")
                except Exception as e:
                    logger.error(f"Failed to parse flag update: {e}")
        except asyncio.CancelledError:
            # Task was cancelled, exit gracefully
            pass
        except Exception as e:
            logger.error(f"Error in PubSub message handler: {e}")

    async def _add_to_index(self, flag: FeatureFlag) -> None:
        """
        Add a flag to the indexes.
        """
        # Add to namespace index
        namespace_key = self._get_namespace_index_key(flag.namespace)
        await self.redis.sadd(namespace_key, flag.key)

        # Add to tag indexes
        for tag_key, tag_value in flag.tags.items():
            tag_index_key = self._get_tag_index_key(flag.namespace, tag_key, tag_value)
            flag_key = self._get_flag_key(flag.key, flag.namespace)
            await self.redis.sadd(tag_index_key, flag_key)

    async def _remove_from_index(self, flag: FeatureFlag) -> None:
        """
        Remove a flag from the indexes.
        """
        # Remove from namespace index
        namespace_key = self._get_namespace_index_key(flag.namespace)
        await self.redis.srem(namespace_key, flag.key)

        # Remove from tag indexes
        for tag_key, tag_value in flag.tags.items():
            tag_index_key = self._get_tag_index_key(flag.namespace, tag_key, tag_value)
            flag_key = self._get_flag_key(flag.key, flag.namespace)
            await self.redis.srem(tag_index_key, flag_key)

    async def _publish_update(self, flag: FeatureFlag) -> None:
        """
        Publish a flag update.
        """
        # Serialize the flag
        flag_dict = flag.dict()
        data = json.dumps(flag_dict)

        # Publish the update
        await self.redis.publish("feature-flag-updates", data)

    def _get_flag_key(self, key: str, namespace: str) -> str:
        """
        Get the Redis key for a flag.
        """
        return f"feature-flag:{namespace}:{key}"

    def _get_namespace_key(self, namespace: str) -> str:
        """
        Get the Redis key prefix for a namespace.
        """
        return f"feature-flag:{namespace}"

    def _get_namespace_index_key(self, namespace: str) -> str:
        """
        Get the Redis key for a namespace index.
        """
        return f"feature-flag:index:namespace:{namespace}"

    def _get_tag_index_key(self, namespace: str, tag_key: str, tag_value: str) -> str:
        """
        Get the Redis key for a tag index.
        """
        return f"feature-flag:index:tag:{namespace}:{tag_key}:{tag_value}"

